### PR TITLE
Proposed onAfterRenderModules event in modules rendering

### PR DESCRIPTION
**Pull request by kanich**
Pull Request for Issue # .
#### Summary of Changes

This PR replaces PR #3105 which was based on outdated staging repository (over 2 years old). Update to current staging repository moved proposed patch from old J3.4 /libraries/joomla/document/html/renderer/modules.php to
/libraries/joomla/document/renderer/html/modules.php and test for this patch is located in /tests/unit/suites/libraries/joomla/document/renderer/html/JDocumentRendererHtmlModulesTest.php (to reflect new directory structure).
Updated patch is in branch 'patch-2' (kanich/joomla-cms)
#### Testing Instructions

Use Joomla unit test for running of /unit/suites/libraries/joomla/document/renderer/html/JDocumentRendererHtmlModulesTest.php. Successfull test run proves correct operation of JDocumentRendererHtmlModules class with proposed change - trigger of onAfterRenderModules event.

…t test

### CODE REVIEW

**Comment by chrisdavenport** [comment]
Please fix code style issues.  Thanks.

**Comment by wojsmol** [comment]
@kanich Please see https://github.com/kanich/joomla-cms/pull/1

**Comment by zero-24** [comment]
I have just send you a CS fix for the unit test. Please ping me if you get it in so i can do a quick test.

**Comment by zero-24** [comment]
I have tested this item :white_check_mark: successfully on bb5068a9756e7bbf7c930a68660bffe32d31ba03<br /><br />It works but the final decision is up to the PLT. Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10113">issues.joomla.org/joomla-cms/10113</a>.</sub>

**Comment by roland-d** [comment]
I have tested this item :white_check_mark: successfully on bb5068a9756e7bbf7c930a68660bffe32d31ba03<br /><br />The test case runs successfully. The code looks good and works as well, I can modify the module buffers.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10113">issues.joomla.org/joomla-cms/10113</a>.</sub>

**Comment by zero-24** [comment]
@roland-d @wilsonge @rdeutz so we can set this RTC or should we wait for a PLT decision?

**Comment by roland-d** [comment]
As far as I am concerned this is RTC for 3.7.

File: libraries/joomla/document/renderer/html/modules.php [hunk 1]
```diff
@@ -50,6 +50,8 @@ public function render($position, $params = array(), $content = null)
 			$buffer .= $moduleHtml;
 		}
 
+		JEventDispatcher::getInstance()->trigger('onAfterRenderModules', array(&$buffer, &$params));
```
Comment (wilsonge) [inline libraries/joomla/document/renderer/html/modules.php]: Does `$buffer` need to be by reference. For caching I can't see why you need to be able to modify html output of modules

**Comment by kanich** [comment]
@wilsonge Yes both array parameters have to passed by reference to be able use these parameters in event processing code. See also : 
Adds an onAfterRenderModule trigger #8296
https://github.com/joomla/joomla-cms/pull/8296/commits/abfc4eea1a7243e2e98e935581f51832d27e6842

For page caching where cached page often contains several modules is this event used for getting latest module content also in case when the page content is still valid (non expired) in page cache. It is used e.g. for refreshing of ecommerce shopping cart on cached product page.

**Pull request merged**